csrc/executor.cpp

```diff
@@ -521,6 +524,178 @@ at::Tensor inferAndAllocOutput(
   return inferAndAlloc(tv, sizes, expr_eval, expand_map, options, zero_init);
 }
 
+std::vector<int64_t> getContiguousStrides(
+    const std::vector<int64_t>& sizes,
+    const std::vector<bool>& expand_flags) {
+  TORCH_INTERNAL_ASSERT(sizes.size() == expand_flags.size());
+
+  std::vector<int64_t> strides(sizes.size());
+  int64_t cur_stride = 1;
+  for (auto i = sizes.size(); i > 0; --i) {
+    auto size = sizes.at(i - 1);
+    TORCH_INTERNAL_ASSERT(
+        size >= 0,
+        "Positive size is assumed non-negative but received: ",
+        size);
+
+    int64_t stride = cur_stride;
+
+    // If expanded, stride is 0
+    if (expand_flags.at(i - 1)) {
+      stride = 0;
+    } else if (size == 0) {
+      // If the size is 0, the stride is 1
+      stride = 1;
+    } else {
+      cur_stride *= size;
+    }
+
+    strides.at(i - 1) = stride;
+  }
+
+  return strides;
+}
+
+// Infer the size and stride of each dimension
+std::pair<std::vector<int64_t>, std::vector<int64_t>> inferShape(
+    const TensorView* tv,
+    std::vector<Val*> symbolic_sizes,
+    std::vector<bool> expand_flags,
+    ExpressionEvaluator& expr_eval) {
+  FUSER_PERF_SCOPE("inferShape");
+
+  // Allocate should be provided for intermediates. We just need to
+  // grab a chunk of memory of the size dicatated by
+  // Allocate::shape(). Fusion outputs do not come with Allocate and
+  // need to be allocated while taking expanded broadcasts into
+  // account.
+
+  std::vector<int64_t> concrete_sizes(symbolic_sizes.size(), 0);
+
+  for (const auto i : c10::irange(symbolic_sizes.size())) {
+    auto symbolic_size = symbolic_sizes.at(i);
+    const auto inferred_val = expr_eval.evaluate(symbolic_size);
+    TORCH_INTERNAL_ASSERT(
+        inferred_val.has_value(),
+        "Could not launch kernel as program could not infer ",
+        symbolic_size->toInlineString(),
+        "(",
+        symbolic_size->toString(),
+        ") for the buffer ",
+        tv->toString());
+
+    auto concrete_size = inferred_val->as<int64_t>();
+    concrete_sizes.at(i) = concrete_size;
+  }
+
+  auto strides = getContiguousStrides(concrete_sizes, expand_flags);
+
+  return {concrete_sizes, strides};
+}
+
+// Infer the shape of an intemediate tensor using kir::Allocate
+std::pair<std::vector<int64_t>, std::vector<int64_t>> inferShapeOfIntermediate(
+    const TensorView* tv,
+    const kir::Allocate* alloc,
+    ExpressionEvaluator& expr_eval) {
+  // Allocate should be provided for intermediates. We just need to
+  // grab a chunk of memory of the size dicatated by
+  // Allocate::shape().
+  TORCH_INTERNAL_ASSERT(alloc != nullptr);
+
+  const auto& symbolic_sizes = alloc->shape();
+  const auto expand_flags = std::vector<bool>(symbolic_sizes.size(), false);
```

Comment:
I know we have already discussed this. Can we just put a note here that we are not expecting any expanded dimensions on buffer?